src/evict/evict_lru.c

```diff
@@ -293,31 +303,44 @@ __wt_evict_thread_run(WT_SESSION_IMPL *session, WT_THREAD *thread)
 		__wt_epoch(session, &cache->stuck_ts);
 #endif
 
-	while (F_ISSET(conn, WT_CONN_EVICTION_RUN) &&
-	    F_ISSET(thread, WT_THREAD_RUN)) {
-		if (conn->evict_server_running &&
-		    __wt_spin_trylock(session, &cache->evict_pass_lock) == 0) {
-			/*
-			 * Cannot use WT_WITH_PASS_LOCK because this is a try
-			 * lock.  Fix when that is supported.  We set the flag
-			 * on both sessions because we may call clear_walk when
-			 * we are walking with the walk session, locked.
-			 */
-			F_SET(session, WT_SESSION_LOCKED_PASS);
-			F_SET(cache->walk_session, WT_SESSION_LOCKED_PASS);
-			ret = __evict_server(session, &did_work);
-			F_CLR(cache->walk_session, WT_SESSION_LOCKED_PASS);
-			F_CLR(session, WT_SESSION_LOCKED_PASS);
-			__wt_spin_unlock(session, &cache->evict_pass_lock);
-			WT_ERR(ret);
-			__wt_verbose(session, WT_VERB_EVICTSERVER, "sleeping");
-
-			/* Don't rely on signals: check periodically. */
-			__wt_cond_auto_wait(
-			    session, cache->evict_cond, did_work, NULL);
-			__wt_verbose(session, WT_VERB_EVICTSERVER, "waking");
+	for (;;) {
+		if (!F_ISSET(thread, WT_THREAD_RUN))
+			break;
+		if (F_ISSET(thread, WT_THREAD_ACTIVE)) {
```

Comment:
Yes, it hides one set of code looking at the thread state for another.  The eviction code has thread shutdown code that it needs to execute.  Since in this style, the eviction code function returns after one pass we now need a shutdown function to call in addition to the run function.  I'll code that up and we can see whether or not it is too gross.